### PR TITLE
Feat: slippage protection for reserve ratio

### PR DESCRIPTION
**Pull request by baroooo**
### Description

Adds slippage protection to `updateRatioForReward()`

### Other changes

Simplifies the new ratio calculation formula

### Tested

Unit tests

### Related issues

- Fixes [#594](https://github.com/mento-protocol/mento-general/issues/594)

### CODE REVIEW

File: contracts/goodDollar/GoodDollarExchangeProvider.sol [hunk 1]
```diff
@@ -190,21 +190,31 @@ contract GoodDollarExchangeProvider is IGoodDollarExchangeProvider, BancorExchan
...
+    uint256 newScaledRatio = unwrap(numerator.div(denominator));
+
+    uint32 newRatioUint = uint32(newScaledRatio / 1e10);
+
+    require(newRatioUint > 0, "New ratio must be greater than 0");
 
-    UD60x18 numerator = wrap(exchange.reserveBalance);
-    UD60x18 denominator = wrap(exchange.tokenSupply + rewardScaled).mul(wrap(currentPriceScaled));
-    uint256 newRatioScaled = unwrap(numerator.div(denominator));
+    uint256 allowedSlippage = (exchange.reserveRatio * maxSlippagePercentage) / 100;
```
Comment (baroooo) [inline contracts/goodDollar/GoodDollarExchangeProvider.sol:213]: is percentage good enough? should we use BPS instead to make it more flexible?

File: contracts/goodDollar/GoodDollarExpansionController.sol [hunk 2]
```diff
@@ -190,12 +190,24 @@ contract GoodDollarExpansionController is IGoodDollarExpansionController, Pausab
...
+  /// @inheritdoc IGoodDollarExpansionController
+  function mintRewardFromReserveRatio(
+    bytes32 exchangeId,
+    address to,
+    uint256 amount,
+    uint256 maxSlippagePercentage
+  ) public onlyAvatar {
     require(to != address(0), "Recipient address must be set");
     require(amount > 0, "Amount must be greater than 0");
+    require(maxSlippagePercentage <= 100, "Max slippage percentage cannot be greater than 100%");
```
Comment (philbow61) [inline contracts/goodDollar/GoodDollarExpansionController.sol:206]: should we check for 0 as well. For super small rewards the call could still go through but generally speaking, the ratio should change looking at the function name in GoodDollarExchangeProvider 😁

File: contracts/goodDollar/GoodDollarExchangeProvider.sol (hunk 1 above)
Comment (philbow61) [reply contracts/goodDollar/GoodDollarExchangeProvider.sol:213]: yes why not we could use maxWeight(1e8) as 100%

**Review by philbow61 — APPROVED** [review]
LGTM left 2 samll comments ✌️

File: contracts/goodDollar/GoodDollarExpansionController.sol (hunk 2 above)
Comment (philbow61) [reply contracts/goodDollar/GoodDollarExpansionController.sol]: haha true 🤝

**Pull request merged**